src/hooks/useViewportColumns.ts

```diff
@@ -1,254 +1,83 @@
+  columns: readonly CalculatedColumn<R, SR>[];
+  colSpanColumns: readonly CalculatedColumn<R, SR>[];
+  rows: readonly (R | GroupRow<R>)[];
+  summaryRows: readonly SR[] | undefined;
+  colOverscanStartIdx: number;
+  colOverscanEndIdx: number;
+  rowOverscanStartIdx: number;
+  rowOverscanEndIdx: number;
+  enableFilterRow: boolean;
+  isGroupRow: (row: R | GroupRow<R>) => row is GroupRow<R>;
 }
 
 export function useViewportColumns<R, SR>({
-  rawColumns,
-  columnWidths,
-  viewportWidth,
-  scrollLeft,
-  defaultColumnOptions,
-  rawGroupBy,
-  enableVirtualization
+  columns,
+  colSpanColumns,
+  rows,
+  summaryRows,
+  colOverscanStartIdx,
+  colOverscanEndIdx,
+  rowOverscanStartIdx,
+  rowOverscanEndIdx,
+  enableFilterRow,
+  isGroupRow
 }: ViewportColumnsArgs<R, SR>) {
-  const minColumnWidth = defaultColumnOptions?.minWidth ?? 80;
-  const defaultFormatter = defaultColumnOptions?.formatter ?? ValueFormatter;
-  const defaultSortable = defaultColumnOptions?.sortable ?? false;
-  const defaultResizable = defaultColumnOptions?.resizable ?? false;
-
-  const { columns, lastFrozenColumnIndex, groupBy } = useMemo(() => {
-    // Filter rawGroupBy and ignore keys that do not match the columns prop
-    const groupBy: string[] = [];
-    let lastFrozenColumnIndex = -1;
-
-    const columns = rawColumns.map(rawColumn => {
-      const rowGroup = rawGroupBy?.includes(rawColumn.key) ?? false;
-      // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
-      const frozen = rowGroup || rawColumn.frozen || false;
-
-      const column: CalculatedColumn<R, SR> = {
-        ...rawColumn,
-        idx: 0,
-        frozen,
-        isLastFrozenColumn: false,
-        rowGroup,
-        sortable: rawColumn.sortable ?? defaultSortable,
-        resizable: rawColumn.resizable ?? defaultResizable,
-        formatter: rawColumn.formatter ?? defaultFormatter
-      };
-
-      if (rowGroup) {
-        column.groupFormatter ??= ToggleGroupFormatter;
-      }
-
-      if (frozen) {
-        lastFrozenColumnIndex++;
-      }
-
-      return column;
-    });
-
-    columns.sort(({ key: aKey, frozen: frozenA }, { key: bKey, frozen: frozenB }) => {
-      // Sort select column first:
-      if (aKey === SELECT_COLUMN_KEY) return -1;
-      if (bKey === SELECT_COLUMN_KEY) return 1;
-
-      // Sort grouped columns second, following the groupBy order:
-      if (rawGroupBy?.includes(aKey)) {
-        if (rawGroupBy.includes(bKey)) {
-          return rawGroupBy.indexOf(aKey) - rawGroupBy.indexOf(bKey);
-        }
-        return -1;
+  // find the column that spans over a column within the visible columns range and adjust colOverscanStartIdx
+  const startIdx = useMemo(() => {
+    let startIdx = colOverscanStartIdx;
+
+    const updateStartIdx = (colIdx: number, colSpan: number | undefined) => {
+      if (!colSpan) return;
+      const newStartIdx = colIdx + colSpan;
+      if (newStartIdx > colOverscanStartIdx && colIdx < startIdx) {
```

Comment:
`newStartIdx` is a confusing variable name here IMO.
```suggestion
      if ((colIdx + colSpan) > colOverscanStartIdx && colIdx < startIdx) {
```